src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,505 @@ g.test('inputs')
+}
+
+/**
+ * Checks that the subgroup size and ballot buffers are consistent.
+ *
+ * This function assumes subgroups will be filled as much as possible with
+ * up to 1 partial subgroup at the end of each workgroup.
+ *
+ * @param data The subgroup_size buffer
+ * @param compare The ballot buffer
+ * @param min The minimum subgroup size allowed
+ * @param max The maximum subgroup size allowed
+ * @param invocations The number of invocations in a workgroup
+ */
+function checkSubgroupSizeConsistency(
+  data: Uint32Array,
+  compare: Uint32Array,
+  min: number,
+  max: number,
+  invocations: number
+): Error | undefined {
+  const subgroupSize = data[0];
+  if (popcount(subgroupSize) !== 1) {
+    return new Error(`Subgroup size '${subgroupSize}' is not a power of two`);
+  }
+  if (subgroupSize < min) {
+    return new Error(`Subgroup size '${subgroupSize}' is less than minimum '${min}'`);
+  }
+  if (max < subgroupSize) {
+    return new Error(`Subgroup size '${subgroupSize}' is greater than maximum '${max}'`);
+  }
+
+  // Check that remaining invocations record a consistent subgroup size.
+  for (let i = 1; i < data.length; i++) {
+    if (data[i] !== subgroupSize) {
+      return new Error(
+        ErrorMsg(`Invocation ${i}: subgroup size inconsistency`, data[i], subgroupSize)
+      );
+    }
+  }
+
+  // Assumes workgroups are divided such that there is a single partial subgroup if any.
+  const partialExpected = invocations % subgroupSize;
+  let fullExpected = 0;
+  if (subgroupSize <= invocations) {
+    fullExpected = Math.floor(invocations / subgroupSize) * subgroupSize;
+  }
+  let fullCount = 0;
+  let partialCount = 0;
+  for (let i = 0; i < compare.length; i++) {
+    if (i % invocations === 0) {
+      if (i !== 0) {
+        // Check intermediate workgroup counts.
+        if (fullCount !== fullExpected) {
+          return new Error(
+            ErrorMsg(`Unexpected number of full invocations`, fullCount, fullExpected)
+          );
+        }
+        if (partialCount !== partialExpected) {
+          return new Error(
+            ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialExpected)
+          );
+        }
+      }
+
+      // Reset for new workgroup.
+      fullCount = 0;
+      partialCount = 0;
+    }
+
+    if (compare[i] === subgroupSize) {
+      fullCount++;
+    } else {
+      partialCount++;
+
+      // Check that all partial invocations are consistent.
+      if (compare[i] !== partialExpected) {
+        return new Error(ErrorMsg(`Partial subgroup size incorrect`, compare[i], partialExpected));
+      }
+    }
+  }
+
+  // Check final workgroup counts.
+  if (fullCount !== fullExpected) {
+    return new Error(ErrorMsg(`Unexpected number of full invocations`, fullCount, fullExpected));
+  }
+  if (partialCount !== partialExpected) {
+    return new Error(
+      ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialExpected)
+    );
+  }
```

Comment:
Maybe change the workgroup milestone condition to simplify a little?
``` suggestion
  for (let i = 0; i < compare.length; i++) {
    if (compare[i] === subgroupSize) {
      fullCount++;
    } else {
      partialCount++;
      // Check that all partial invocations are consistent.
      if (compare[i] !== partialExpected) {
        return new Error(ErrorMsg(`Partial subgroup size incorrect`, compare[i], partialExpected));
      }
    }

    // Validate the counters at the end of each workgroup.
    if ((i+1) % invocations === 0) {
      // Check intermediate workgroup counts.
      if (fullCount !== fullExpected) {
        return new Error(
          ErrorMsg(`Unexpected number of full invocations`, fullCount, fullExpected)
        );
      }
      if (partialCount !== partialExpected) {
        return new Error(
          ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialExpected)
        );
      }

      // Reset counters for next workgroup.
      fullCount = 0;
      partialCount = 0;
    }
  }
```

Reply:
I thought should also validate/assert `compare.length % invocations === 0` in this case.

Reply:
I ended up loosening this check.